cmd/demo/node_setup.go

```diff
@@ -103,63 +105,125 @@ func (n *node) setup() error {
 	return nil
 }
 
-// setupContracts reads from the config file whether the node should deploy or use
-// existing contract addresses.
-func (n *node) setupContracts() (err error) {
-	var adjAddr, assAddr common.Address
+func getTxContext() (context.Context, context.CancelFunc) {
+	return context.WithTimeout(context.Background(), config.Chain.TxTimeout)
+}
 
-	if config.Chain.Adjudicator == "deploy" {
-		adjAddr, err = deployAdjudicator(n.cb)
-		if err != nil {
-			return
-		}
-		fmt.Println("💭 Adjudicator contract deployed")
-	} else {
-		tmpAdj, err := strToAddress(config.Chain.Adjudicator)
-		if err != nil {
-			return err
-		}
-		adjAddr = ewallet.AsEthAddr(tmpAdj)
+func (n *node) validateAdjudicatorContract(adjAddr common.Address) error {
+	ctx, cancel := getTxContext()
+	defer cancel()
+	return echannel.ValidateAdjudicator(ctx, n.cb, adjAddr)
+}
 
-		ctx, cancel := context.WithTimeout(context.Background(), config.Chain.TxTimeout)
-		defer cancel()
-		if err := echannel.ValidateAdjudicator(ctx, n.cb, adjAddr); err != nil {
-			return errors.WithMessage(err, "validating adjudicator contract")
-		}
+func (n *node) validateAssetHolderContract(assAddr common.Address, adjAddr common.Address) error {
+	ctx, cancel := getTxContext()
+	defer cancel()
+	return echannel.ValidateAssetHolderETH(ctx, n.cb, assAddr, adjAddr)
+}
+
+func stringToAddress(s string) (common.Address, error) {
+	var err error
+	var walletAddr wallet.Address
 
-		fmt.Println("💭 Adjudicator contract validated")
+	if walletAddr, err = strToAddress(s); err != nil {
+		return common.Address{}, err
 	}
 
-	if config.Chain.Assetholder == "deploy" {
-		assAddr, err = deployAsset(n.cb, adjAddr)
-		if err != nil {
-			return
-		}
-		fmt.Println("💭 Asset holder contract deployed")
-	} else {
-		tmpAsset, err := strToAddress(config.Chain.Assetholder)
-		if err != nil {
-			return err
+	return ewallet.AsEthAddr(walletAddr), nil
+}
+
+func (n *node) setupAdjudicatorContract() error {
+	// config options: "[onChainAddress]", "deploy", "[onChainAddress]|deploy"
+
+	tokens := strings.Split(config.Chain.Adjudicator, "|")
```

Comment:
If we use this config "hack", it should probably be parsed in the config setup (`cmd/demo.SetConfig`). I think it would be the cleanest to add two options to the `chainConfig` and then set them in `SetConfig`. Namely: `DeployAdjudicator` and `DeployAssetholder`, which can then be used in the functions here without having to parse the strings here. Same for the `setupAssetHolderContract`.  
  
Then please also fix the possible combinations when you parse the string, because right now `deploy|deploy|0x232…` would be valid.

Reply:
> If we use this config "hack", it should probably be parsed in the config setup (`cmd/demo.SetConfig`). I think it would be the cleanest to add two options to the `chainConfig` and then set them in `SetConfig`. Namely: `DeployAdjudicator` and `DeployAssetholder`, which can then be used in the functions here without having to parse the strings here. Same for the `setupAssetHolderContract`.

Ok. How about just adding one config option`deployOnValidationFailure`, which means "if any contract validation fails, try to deploy the contract"? I don't see a case in the demo where one needs separate control over each contract deployment.

> Then please also fix the possible combinations when you parse the string, because right now `deploy|deploy|0x232…` would be valid.

This was actually intended, but becomes obsolete with your current suggestion.